### PR TITLE
Initial implementation of GitLab statistics

### PR DESCRIPTION
**Pull request by nE0sIghT**
Issue: https://github.com/pypi/warehouse/issues/12789

### CODE REVIEW

**Comment by di** [comment]
Thanks for the PR! It looks great, but I'm going to hold off on fully reviewing/merging this until #8635 is resolved for GitLab URLs, so we don't run into the same issue as we have for GitHub.

**Comment by nE0sIghT** [comment]
#16205 looks good! I will follow it and work on Gitlab part if needed

**Review by firefly9966 — APPROVED** [review]

**Review by firefly9966 — APPROVED** [review]

**Comment by nE0sIghT** [comment]
Hello!

Do I need to rebase this PR or either merge `main` branch here?

**Comment by di** [comment]
@nE0sIghT Hang tight, I suspect the next part of https://github.com/pypi/warehouse/issues/8635 that we will resolve will be verifying the GitHub statistics, after which we can get this up to date. Thanks for your patience!

**Comment by nE0sIghT** [comment]
Could we proceed with this PR? I can rebase or merge main branch here

**Comment by di** [comment]
Sorry, I meant to comment here that this is unblocked now that https://github.com/pypi/warehouse/pull/16532 is merged. Please feel free to proceed!

**Comment by nE0sIghT** [comment]
Thanks! Should I merge or rebase?

**Comment by di** [comment]
Either is fine, we do squash merging so it doesn't make much difference.

**Comment by nE0sIghT** [comment]
I rebased PR on top of the main branch, added unit tests and tested locally

image

File: warehouse/csp.py [hunk 1]
```diff
@@ -97,6 +97,7 @@ def _connect_src_settings(config) -> list:
         SELF,
         "https://api.github.com/repos/",
         "https://api.github.com/search/issues",
+        "https://gitlab.com",
```
Comment (di) [inline warehouse/csp.py:100]:
```suggestion
        "https://gitlab.com/api",
```

File: warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          forks_url: project.web_url + "/-/forks",
+          MRs: merge_requests_count,
+          MRs_url: project.web_url + "/-/merge_requests",
+        };
+        this.gitlabRepoInfoOutlets.forEach((outlet) =>
+          outlet.updateStats(stats)
+        );
+      })
+      // swallow errors, we don't want to show them to the user
+      .catch(() => {});
```
Comment (di) [inline warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js:80]: We don't want to show them to the user, but would probably be helpful to at least log them to the console here.

File: warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }),
+        fetchParams
+      ).then((response) => {
+        if(response.ok) {
+          return response.json().then(
+            (data) => Object.keys(data).length ? response.headers.get("X-Total-Pages") : 0
+          );
+        }
+
+        return 0;
```
Comment (di) [inline warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js:41]: Similarly would be nice to log any errors here to the console.

File: warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { Controller } from "@hotwired/stimulus";
+
+export default class extends Controller {
+  static targets = [
+    "forksCount",
+    "forksUrl",
+    "openIssuesCount",
+    "openIssuesUrl",
+    "stargazersCount",
+    "stargazersUrl",
```
Comment (di) [inline warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js:23]: "stargazers" seems to be a github-ism, we should just call this "followers" I think.

File: warehouse/templates/includes/packaging/project-data.html [hunk 5]
```diff
@@ -92,6 +92,53 @@ <h6>{% trans %}GitHub Statistics{% endtrans %}</h6>
...
+          <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
+          <strong>{% trans %}Open issues:{% endtrans %}</strong>
+          <span data-gitlab-repo-info-target="openIssuesCount"></span>
+        </a>
+      </li>
+      <li>
+        <a class="vertical-tabs__tab vertical-tabs__tab--with-icon vertical-tabs__tab--condensed"
+           data-gitlab-repo-info-target="openMRsUrl" rel="noopener">
+          <i class="fa fa-code-pull-request" aria-hidden="true"></i>
+          <strong>{% trans %}Open MRs:{% endtrans %}</strong>
```
Comment (di) [inline warehouse/templates/includes/packaging/project-data.html:134]:
```suggestion
          <strong>{% trans %}Open merge requests:{% endtrans %}</strong>
```

File: warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js (hunk 4 above)
Comment (nE0sIghT) [reply warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js]: I used "starrers" term here as in the Gitlab's UI and API

**Comment by nE0sIghT** [comment]
Thanks for review! I pushed new commits that should address mentioned issues.

File: warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js (hunk 4 above)
Comment (di) [reply warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js]: Ah, perfect, ty!

**Review by di — APPROVED** [review]
Thanks!

**Comment by di** [comment]
@nE0sIghT Can you resolve the merge conflicts here? Thanks

**Comment by nE0sIghT** [comment]
> Can you resolve the merge conflicts here?

It's done

**Comment by di** [comment]
Sorry, we have been sorting out some issues there recently :) Yes please, should be the last time!

**Pull request merged**